From 0fce0c80fd6210695e333ad512cf6b05a8012a2a Mon Sep 17 00:00:00 2001
From: "Enrico Weigelt, metux IT consult" <info@metux.net>
Date: Thu, 26 Jun 2025 13:46:08 +0200
Subject: [PATCH] xfree86: compat: re-add GEInitEvent() for proprietary nvidia
 driver

Yet another very internal function that the proprietary Nvidia driver
is using for unknown reasons. NVidia really needs a separate function
for just for some trivial struct initialization and don't manage to
add three simple lines to their code, so we have to make an extra
function for them.

Signed-off-by: Enrico Weigelt, metux IT consult <info@metux.net>
---
 hw/xfree86/compat/geeventinit.c | 28 ++++++++++++++++++++++++++++
 hw/xfree86/compat/meson.build   |  1 +
 2 files changed, 29 insertions(+)
 create mode 100644 hw/xfree86/compat/geeventinit.c

diff --git a/hw/xfree86/compat/geeventinit.c b/hw/xfree86/compat/geeventinit.c
new file mode 100644
index 0000000000..9069ba08ed
--- /dev/null
+++ b/hw/xfree86/compat/geeventinit.c
@@ -0,0 +1,28 @@
+#include <dix-config.h>
+
+#include <X11/Xfuncproto.h>
+#include <X11/Xproto.h>
+
+#include "os/osdep.h"
+
+#include "xf86_compat.h"
+
+/*
+ * needed for NVidia proprietary driver 340.x versions
+ *
+ * they really need special functions for trivial struct initialization :p
+ *
+ * this function had been obsolete and removed long ago, but NVidia folks
+ * still didn't do basic maintenance and fixed their driver
+ */
+
+_X_EXPORT void GEInitEvent(xGenericEvent *ev, int extension);
+
+void GEInitEvent(xGenericEvent *ev, int extension)
+{
+    xf86NVidiaBugObsoleteFunc("GEInitEvent()");
+
+    ev->type = GenericEvent;
+    ev->extension = extension;
+    ev->length = 0;
+}
diff --git a/hw/xfree86/compat/meson.build b/hw/xfree86/compat/meson.build
index 79157c9497..2e34856152 100644
--- a/hw/xfree86/compat/meson.build
+++ b/hw/xfree86/compat/meson.build
@@ -8,6 +8,7 @@ srcs_xorg_compat = [
 
 if get_option('legacy_nvidia_340x')
     srcs_xorg_compat += 'timercheck.c'
+    srcs_xorg_compat += 'geeventinit.c'
 endif
 
 xorg_compat = static_library('xorg_compat',
